File: airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/ParsedCatalog.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.base.destination.typing_deduping;

import io.airbyte.protocol.models.v0.AirbyteStreamNameNamespacePair;
import java.util.List;

public record ParsedCatalog(List<StreamConfig> streams) {

  public StreamConfig getStream(AirbyteStreamNameNamespacePair streamId) {
    return getStream(streamId.getNamespace(), streamId.getName());
  }

  public StreamConfig getStream(StreamId streamId) {
    return getStream(streamId.originalNamespace(), streamId.originalName());
  }

  public StreamConfig getStream(String originalNamespace, String originalName) {
    return streams.stream()
        .filter(s -> s.id().originalNamespace().equals(originalNamespace) && s.id().originalName().equals(originalName))
        .findFirst()
        .orElseThrow(() -> new IllegalArgumentException(String.format(
            "Could not find stream %s.%s out of streams %s",
            originalNamespace,
            originalName,
            streams.stream().map(stream -> stream.id().originalNamespace() + "." + stream.id().originalName()).toList())));
  }

}

File: airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/kotlin/io/airbyte/integrations/base/destination/typing_deduping/ParsedCatalog.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.integrations.base.destination.typing_deduping

import io.airbyte.protocol.models.v0.AirbyteStreamNameNamespacePair

class ParsedCatalog(val streams: List<StreamConfig>) {
    fun getStream(streamId: AirbyteStreamNameNamespacePair): StreamConfig {
        return getStream(streamId.namespace, streamId.name)
    }

    fun getStream(streamId: StreamId): StreamConfig {
        return getStream(streamId.originalNamespace, streamId.originalName)
    }

    fun getStream(originalNamespace: String?, originalName: String?): StreamConfig {
        return streams.firstOrNull { s: StreamConfig ->
            s.id.originalNamespace == originalNamespace && s.id.originalName == originalName
        }
            ?: throw IllegalArgumentException(
                String.format(
                    "Could not find stream %s.%s out of streams %s",
                    originalNamespace,
                    originalName,
                    streams
                        .map { stream: StreamConfig ->
                            stream.id.originalNamespace + "." + stream.id.originalName
                        }
                        .toList()
                )
            )
    }
}
